File: frontend/ianus-core/LicenseValidation.ts
import { ILicense } from "./License";
import { LicenseValidationResult } from "./LicenseValidationResult";

// from https://developers.google.com/web/updates/2012/06/How-to-convert-ArrayBuffer-to-and-from-String
const str2ab = (str: string) => {
    const buf = new ArrayBuffer(str.length);
    const bufView = new Uint8Array(buf);
    for (let i = 0, strLen = str.length; i < strLen; i++) {
      bufView[i] = str.charCodeAt(i);
    }
    return buf;
};
 
const importRsaKey = (pem: string) => {
    // base64 decode the string to get the binary data
    const binaryDerString = window.atob(pem.replace("-----BEGIN PUBLIC KEY-----", "").replace("-----END PUBLIC KEY-----","").trim());
    // convert from a binary string to an ArrayBuffer
    const binaryDer = str2ab(binaryDerString);
  
    return window.crypto.subtle.importKey(
      "spki",
      binaryDer,
      {
        name: "RSASSA-PKCS1-v1_5",
        hash: "SHA-256",
      },
      true,
      ["verify"]
    );
};

const validateClaims = (validIsvId: string, validProductId: string, organizationId: string, license: ILicense): [boolean, string] => {
    if (!license) {
        return [false, "No license passed!"];
    }

    if (!license.env || !license.env.length || !license.aud || !license.iss) {
        return [false, "Incomplete license!"];
    }

    const validIssuer = `https://www.ianusguard.com/api/public/products/${validProductId}`;

    if (license.iss.toLowerCase() !== validIssuer.toLowerCase()) {
        return [false, `Invalid license issuer: Issuer must be '${validIssuer}'`];
    }

    const validAudience = "ianusguard";

    if (license.aud !== validAudience) {
        return [false, `Invalid license audience: Audience must be '${validIsvId}'`];
    }

    if (license.isv !== validIsvId) {
        return [false, `Invalid license ISV: ISV must be '${validIsvId}'`];
    }

    if (license.prd !== validProductId) {
        return [false, `Invalid license product: Product must be '${validProductId}'`];
    }

    const formattedOrganizationId = organizationId.replace("{", "").replace("}", "").toLowerCase();
    if (!license.env.includes(formattedOrganizationId)) {
        return [false, `Invalid organization: Your license is not intended for usage in '${formattedOrganizationId}' but for '${JSON.stringify(license.env)}'`];
    }

    // Licenses without exp are considered not expiring
    if (license.exp != null )
    {
        if(isNaN(license.exp)) {
            return [false, "Invalid license expiry: Expiry is not a number"];
        }

        const expiryDate = new Date(license.exp * 1000);
        if (expiryDate < new Date()) {
            return [false, `Invalid license expiry: License expired on '${expiryDate}'`];
        }
    }

    return [true, ""];
};

// Helper for Base64url encoding, from: https://thewoods.blog/base64url/
const base64url_decode = (value: string): ArrayBuffer => {
    const m = value.length % 4;
    return Uint8Array.from(atob(
        value.replace(/-/g, '+')
            .replace(/_/g, '/')
            .padEnd(value.length + (m === 0 ? 0 : 4 - m), '=')
    ), c => c.charCodeAt(0)).buffer
};

const verifySignature = async (key: CryptoKey, dataToVerify: BufferSource, signature: BufferSource) =>
{
    return await window.crypto.subtle.verify(
        "RSASSA-PKCS1-v1_5",
        key,
        signature,
        dataToVerify
    );
}

export const validateLicense = async (validIsvId: string, validProductId: string, organizationId: string, publicKey: string, licenseKey: string | undefined): Promise<LicenseValidationResult> => {
    if (!licenseKey) {
        return {
            isValid: false,
            reason: "No license key passed!"
        };
    }

    try {
        const key = await importRsaKey(publicKey);
        const parts = licenseKey.split('.');

        if (parts.length < 3) {
            return {
                isValid: false,
                reason: "Incorrect license!"
            };
        }

        const [encodedHeaders, encodedClaims, signature] = parts;

        const plainClaims = window.atob(encodedClaims);
        const claimsJson = JSON.parse(plainClaims);

        const [claimsValidationResult, claimsValidationResultMessage] = validateClaims(validIsvId, validProductId, organizationId, claimsJson);

        if (!claimsValidationResult) {
            return {
                isValid: false,
                reason: claimsValidationResultMessage
            };
        }

        const dataToVerify = new TextEncoder().encode(encodedHeaders + "." + encodedClaims);
        const signatureToVerify = base64url_decode(signature);
        const isLicenseSignatureValid = verifySignature(key, dataToVerify, signatureToVerify);

        if (!isLicenseSignatureValid)
        {
            return {
                isValid: false,
                reason: "Invalid license signature: Verification failed!"
            };
        }
        else
        {
            return {
                isValid: true,
                reason: "",
                license: claimsJson
            };
        }
    }
    catch {
        return {
            isValid: false,
            reason: "Oops, something went wrong while validating your license"
        };
    }
};

File: backend/ianus-client/LicenseValidation.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Security.Cryptography;
using Microsoft.Xrm.Sdk;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.OpenSsl;
using Microsoft.Xrm.Sdk.Messages;
using Microsoft.Xrm.Sdk.Query;
using System.Collections.Generic;

namespace Ianua.Ianus.Client
{
    public static class LicenseValidation
    {
        public static RSA ImportRsaPublicKey(string pem)
        {
            var rsaParams = ConvertPemToRsaParameters(pem);

            var rsa = RSA.Create();
            rsa.ImportParameters(rsaParams);

            return rsa;
        }

        private static RSAParameters ConvertPemToRsaParameters(string pem)
        {
            using (var publicKeyTextReader = new StringReader(pem))
            {
                var pemReader = new PemReader(publicKeyTextReader);
                var rsaKeyParameters = (RsaKeyParameters)pemReader.ReadObject();

                var rsaParams = new RSAParameters
                {
                    Modulus = rsaKeyParameters.Modulus.ToByteArrayUnsigned(),
                    Exponent = rsaKeyParameters.Exponent.ToByteArrayUnsigned()
                };

                return rsaParams;
            }
        }

        private static byte[] Base64UrlDecode(string input)
        {
            string base64 = input.Replace('-', '+').Replace('_', '/');

            switch (input.Length % 4)
            {
                case 2: base64 += "=="; break;
                case 3: base64 += "="; break;
            }

            return Convert.FromBase64String(base64);
        }

        private static bool VerifySignature(RSA key, byte[] data, byte[] signature)
        {
            return key.VerifyData(data, signature, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
        }

        private static Guid RetrieveOrganizationId (IOrganizationService service)
        {
            var request = new RetrieveMultipleRequest
            {
                Query = new QueryExpression
                {
                    EntityName = "organization",
                    ColumnSet = new ColumnSet("organizationid"),
                    TopCount = 1
                }
            };

            var results = (RetrieveMultipleResponse) service.Execute(request);

            return results.EntityCollection.Entities.FirstOrDefault()?.Id ?? Guid.Empty;
        }

        public static LicenseValidationResult ValidateClaims
        (
            Guid validIsvId, 
            Guid validProductId, 
            Guid organizationId, 
            LicenseClaims licenseClaims
        )
        {
            if (licenseClaims == null)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "No license passed!"
                };
            }

            if (licenseClaims.Env == null || !licenseClaims.Env.Any() || string.IsNullOrEmpty(licenseClaims.Aud) || string.IsNullOrEmpty(licenseClaims.Iss))
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "Incomplete license!"
                };
            }

            var validIssuer = $"https://www.ianusguard.com/api/public/products/{validProductId}";

            if (!string.Equals(validIssuer, licenseClaims.Iss, StringComparison.InvariantCultureIgnoreCase))
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = $"Invalid license issuer: License must be issued by '{validIssuer}'"
                };
            }

            if (!string.Equals(licenseClaims.Aud, "ianusguard", StringComparison.InvariantCultureIgnoreCase))
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = $"Invalid license audience: License must have audience 'ianusguard'"
                };
            }

            if (licenseClaims.Isv != validIsvId)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = $"Invalid license ISV: License must be issued by '{validIsvId}'"
                };
            }

            if (licenseClaims.Prd != validProductId)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = $"Invalid license product: License product must be '{validProductId}'"
                };
            }

            if (!licenseClaims.Env.Contains(organizationId))
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = $"Invalid organization: Your license is not intended for usage in '{organizationId}' but for '{string.Join(", ", licenseClaims.Env)}'"
                };
            }

            // A license without exp claim is defined to not expire
            if (licenseClaims.Exp != null)
            {
                // Validate expiration date
                var expiryDate = DateTimeOffset.FromUnixTimeSeconds(licenseClaims.Exp.Value).DateTime;
                if (expiryDate < DateTime.UtcNow)
                {
                    return new LicenseValidationResult
                    {
                        IsValid = false,
                        Reason = $"Invalid license expiry: Your license has expired on '{expiryDate}'"
                    };
                }
            }

            return new LicenseValidationResult
            {
                IsValid = true,
                License = new License
                {
                    Claims = licenseClaims
                }
            };
        }

        public static LicenseValidationResult ValidateLicense(Guid validIsvId, Guid validProductId, string publicKeyPem, string licenseKey, IOrganizationService service)
        {
            var key = ImportRsaPublicKey(publicKeyPem);

            if (string.IsNullOrEmpty(licenseKey))
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "No license key set!"
                };
            }

            // Split the license key into parts
            var parts = licenseKey.Split('.');

            if (parts.Length < 3)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "Invalid license format!"
                };
            }

            var encodedHeaders = parts[0];
            var encodedClaims = parts[1];
            var signature = parts[2];

            // Base64 decode the claims
            var plainClaims = Base64UrlDecode(encodedClaims);
            var licenseClaims = JsonSerializer.Deserialize<LicenseClaims>(plainClaims);

            var organizationId = RetrieveOrganizationId(service);

            if (organizationId == Guid.Empty)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "Failed to retrieve dataverse organization unique name!"
                };
            }

            var licenseValidationResult = ValidateClaims(validIsvId, validProductId, organizationId, licenseClaims);

            if (!licenseValidationResult.IsValid)
            {
                return licenseValidationResult;
            }

            // Create the data to verify (headers.claims)
            var dataToVerify = Encoding.UTF8.GetBytes($"{encodedHeaders}.{encodedClaims}");

            // Verify the signature
            var isLicenseSignatureValid = VerifySignature(key, dataToVerify, Base64UrlDecode(signature));

            if (!isLicenseSignatureValid)
            {
                return new LicenseValidationResult
                {
                    IsValid = false,
                    Reason = "Invalid license signature: Verification failed!"
                };
            }
            else
            {
                return licenseValidationResult;
            }
        }
    }
}
